runtime/codefrag.c

```diff
@@ -64,18 +64,18 @@ int caml_register_code_fragment(char * start, char * end,
 
 void caml_remove_code_fragment(struct code_fragment * cf)
 {
-  struct code_fragment * c;
-  int i;
-
   caml_skiplist_remove(&caml_code_fragments_list, (uintnat) cf->code_start);
-  /* Remove cf from the table, maintaining the invariant that
-     if c is the i-th element of the table, c->fragnum is equal to i. */
-  for (i = cf->fragnum; i < caml_code_fragments_table.size - 1; i++) {
-    c = caml_code_fragments_table.contents[i + 1];
-    caml_code_fragments_table.contents[i] = c;
-    c->fragnum = i;
+  /* Currently, caml_remove_code_fragment is used only to remove
+     the fragment most recently inserted, so optimize for this case. */
+  if (cf->fragnum == caml_code_fragments_table.size - 1) {
+    caml_code_fragments_table.size--;
+  } else {
+    /* Just mark the table entry invalid.  Later, we could maintain
+       a free list of empty entries in caml_code_fragments_table
+       so as to reuse them on the next caml_register_code_fragment.
+       Or replace the table by a skiplist. */
```

Comment:
Is there any major complication in using a skiplist? If the code is no more complicated, then let's use the alternative which works better from the beginning (i.e., use a skiplist).

Reply:
The things you make me do...  See commit 0107f395f .

Reply:
I'm honored to make you do these things ;)

Reply:
I had a look at the code : this seems good to me.